Python/specialize.c

```diff
@@ -568,8 +574,47 @@ analyze_descriptor(PyTypeObject *type, PyObject *name, PyObject **descr, int sto
     }
     else {
         if (type->tp_getattro != PyObject_GenericGetAttr) {
-            *descr = NULL;
-            return GETSET_OVERRIDDEN;
+            getattrofunc getattro_slot = type->tp_getattro;
```

Comment:
I wonder if our assumption that `getattro_slot == _Py_slot_tp_getattr_hook` implies that there is a `__getattr__` is true.
It might be that the slot hasn't been updated, even though that seems unlikely.

Could you add an assert that `_PyType_Lookup(type, &_Py_ID(__getattr__)) != NULL` if `has_getattr`?

Reply:
`_Py_slot_tp_getattr_hook` handles this case by overwriting itself back to the simpler `_Py_slot_tp_getattro`. But in this case would it matter? If there's no `__getattr__` isn't that a good thing? It doesn't change the fact that we can just do normal attribute lookup.

Reply:
What are we checking for here?

This is what I think we want to do w.r.t overriding these methods:
* Neither `__getattribute__` nor `__getattr__` are overridden, which we already handle.
* Both are overridden. Rare, don't specialize.
* `__getattribute__` is overridden, `__getattr__` is not, and `__getattribute__` is a Python function, which we add a new specialization for.
* `__getattr__` is overridden, two cases:
   *  We expect a value on the instance. We can treat this as a normal instance lookup, as we de-opt if the value is `NULL`, falling back to the full lookup.
  * No attribute on the instance. This is going to be rare, so don't specialize.

Is that correct?

If so, could you set two booleans `has_getattr` and `has_getattribute`first, then do the case analysis, it would be clearer (to me at least).